### PR TITLE
Adding seedgenerator to the lca package

### PR DESCRIPTION
**Pull request by achuzhoy**
This will allow automating tests against seedgenerator.

### CODE REVIEW

**Review by cdvultur — DISMISSED** [review]
/lgtm

File: pkg/lca/seedgenerator.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+	lcasgv1alpha1 "github.com/openshift-kni/lifecycle-agent/api/seedgenerator/v1alpha1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SeedGenerator provides struct for the seedgenerator object containing connection to
+// the cluster and the seedgenerator definitions.
+type SeedGenerator struct {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:20]:
```suggestion
type SeedGeneratorBuilder struct {
```

File: pkg/lca/seedgenerator.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+		glog.V(100).Infof("The name of the seedgenerator is empty")
+
+		builder.errorMsg = "SeedGenerator name cannot be empty"
+	}
+
+	return &builder
+}
+
+// WithOptionsSeedGenerator creates seedgenerator with generic mutation options.
+func (builder *SeedGenerator) WithOptionsSeedGenerator(options ...SeedGeneratorAdditionalOptions) *SeedGenerator {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:58]:
```suggestion
func (builder *SeedGenerator) WithOptions(options ...SeedGeneratorAdditionalOptions) *SeedGenerator {
```
This is a method do it doesn't have to be unique

File: pkg/lca/seedgenerator.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+		return nil, fmt.Errorf("seedgenerator object %s doesn't exist", name)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// DeleteSeedGenerator removes the existing seedgenerator from a cluster.
+func (builder *SeedGenerator) DeleteSeedGenerator() (*SeedGenerator, error) {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:131]:
```suggestion
func (builder *SeedGenerator) Delete() (*SeedGenerator, error) {
```
This is a method so it doesn't have to be unique

File: pkg/lca/seedgenerator.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+		return builder, fmt.Errorf("can not delete seedgenerator: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// GetSeedGenerator returns seedgenerator object if found.
+func (builder *SeedGenerator) GetSeedGenerator() (*lcasgv1alpha1.SeedGenerator, error) {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:155]:
```suggestion
func (builder *SeedGenerator) Get() (*lcasgv1alpha1.SeedGenerator, error) {
```
This is a method so it doesn't have to be unique

File: pkg/lca/seedgenerator.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	return seedgenerator, err
+}
+
+// ExistsSeedGenerator checks whether the given seedgenerator exists.
+func (builder *SeedGenerator) ExistsSeedGenerator() bool {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:176]:
```suggestion
func (builder *SeedGenerator) Exists() bool {
```
This is a method so it doesn't have to be unique

File: pkg/lca/seedgenerator.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+		builder.Definition.Name)
+
+	var err error
+	builder.Object, err = builder.GetSeedGenerator()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithSeedImageSeedGenerator sets the seed image used by the seedgenerator.
+func (builder *SeedGenerator) WithSeedImageSeedGenerator(
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:191]:
```suggestion
func (builder *SeedGenerator) WithSeedImage(
```
Adding SeedGenerator to methods names is redundant

File: pkg/lca/seedgenerator.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+	glog.V(100).Infof("Setting seed image %s in seedgenerator", seedImage)
+
+	builder.Definition.Spec.SeedImage = seedImage
+
+	return builder
+}
+
+// WithRecertImageSeedGenerator sets the recert image used by the seedgenerator.
+func (builder *SeedGenerator) WithRecertImageSeedGenerator(
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:205]:
```suggestion
func (builder *SeedGenerator) WithRecertImage(
```
Adding SeedGenerator to methods names is redundant

File: pkg/lca/seedgenerator.go [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+	glog.V(100).Infof("Setting recert image %s in seedgenerator", recertImage)
+
+	builder.Definition.Spec.RecertImage = recertImage
+
+	return builder
+}
+
+// WaitUntilCompleteSeedGenerator waits the specified timeout for the seedgenerator to complete
+// actions.
+func (builder *SeedGenerator) WaitUntilCompleteSeedGenerator() (*SeedGenerator, error) {
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:220]:
```suggestion
func (builder *SeedGenerator) WaitUntilComplete() (*SeedGenerator, error) {
```
Adding SeedGenerator to methods names is redundant

Comment (achuzhoy) [reply pkg/lca/seedgenerator.go:220]: Someone suggested we should do it to differentiate between 2 different components of the same package.

Comment (trewest) [reply pkg/lca/seedgenerator.go:220]: That is only for functions

Comment (trewest) [reply pkg/lca/seedgenerator.go:220]: i.e. lca pkg cannot have two functions called Pull

File: pkg/lca/seedgenerator.go (hunk 1 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 3 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 4 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 5 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 7 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 8 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

File: pkg/lca/seedgenerator.go (hunk 2 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go]: Updated. Thanks.

**Comment by achuzhoy** [comment]
cc @mpmaruthu

File: pkg/lca/seedgenerator.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+// WithSeedImage sets the seed image used by the seedgenerator.
+func (builder *SeedGeneratorBuilder) WithSeedImage(
+	seedImage string) *SeedGeneratorBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Setting seed image %s in seedgenerator", seedImage)
+
+	builder.Definition.Spec.SeedImage = seedImage
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:199]: It can be empty?

File: pkg/lca/seedgenerator.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+// WithRecertImage sets the recert image used by the seedgenerator.
+func (builder *SeedGeneratorBuilder) WithRecertImage(
+	recertImage string) *SeedGeneratorBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Setting recert image %s in seedgenerator", recertImage)
+
+	builder.Definition.Spec.RecertImage = recertImage
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:213]: It can be empty?

File: pkg/lca/seedgenerator.go [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+	return builder
+}
+
+// WaitUntilComplete waits the specified timeout for the seedgenerator to complete
```
Comment (trewest) [inline pkg/lca/seedgenerator.go:218]: Should there be a timeout parameter?

File: pkg/lca/seedgenerator.go (hunk 10 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go:213]: yes - never actually set it :)

File: pkg/lca/seedgenerator.go (hunk 9 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go:199]: yes - you can create a seedimage with empty spec.

`oc get seedgenerators.lca.openshift.io seedimage -o yaml
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  creationTimestamp: "2024-01-18T17:05:02Z"
  generation: 1
  name: seedimage
  resourceVersion: "1896527"
  uid: 43fc7e08-68a6-42b2-b97e-64696aec2326
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-01-18T17:05:02Z"
    message: Seed Generation in progress
    observedGeneration: 1
    reason: InProgress
    status: "True"
    type: SeedGenInProgress
  observedGeneration: 1
`

Comment (trewest) [reply pkg/lca/seedgenerator.go:199]: Not empty spec, it should be empty spec.seedImage

File: pkg/lca/seedgenerator.go (hunk 11 above)
Comment (achuzhoy) [reply pkg/lca/seedgenerator.go:218]: added. Thanks.

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**